File: packages/react-native-autoplay/ios/HybridAutoPlay.swift
import CarPlay
import NitroModules

class HybridAutoPlay: HybridHybridAutoPlaySpec {
    private static var listeners = [EventName: [String: () -> Void]]()
    private static var renderStateListeners = [
        String: [(VisibilityState) -> Void]
    ]()
    private static var safeAreaInsetsListeners = [
        String: [(SafeAreaInsets) -> Void]
    ]()

    func addListener(eventType: EventName, callback: @escaping () -> Void)
        throws -> () -> Void
    {
        let uuid = UUID().uuidString
        HybridAutoPlay.listeners[eventType, default: [:]][uuid] = callback

        if eventType == .didconnect && SceneStore.isRootModuleConnected() {
            callback()
        }

        return {
            HybridAutoPlay.listeners[eventType]?.removeValue(forKey: uuid)
        }
    }

    func addListenerRenderState(
        mapTemplateId: String,
        callback: @escaping (VisibilityState) -> Void
    ) throws -> () -> Void {
        if HybridAutoPlay.renderStateListeners[mapTemplateId] != nil {
            HybridAutoPlay.renderStateListeners[mapTemplateId]?.append(callback)
        } else {
            HybridAutoPlay.renderStateListeners[mapTemplateId] = [callback]
        }

        if let state = SceneStore.getState(moduleName: mapTemplateId) {
            callback(state)
        }

        return {
            HybridAutoPlay.renderStateListeners[mapTemplateId]?.removeAll {
                $0 as AnyObject === callback as AnyObject
            }
            if HybridAutoPlay.renderStateListeners[mapTemplateId]?.isEmpty
                ?? false
            {
                HybridAutoPlay.renderStateListeners.removeValue(
                    forKey: mapTemplateId
                )
            }
        }
    }

    func addSafeAreaInsetsListener(
        moduleName: String,
        callback: @escaping (SafeAreaInsets) -> Void
    ) throws -> () -> Void {
        if HybridAutoPlay.safeAreaInsetsListeners[moduleName] != nil {
            HybridAutoPlay.safeAreaInsetsListeners[moduleName]?.append(callback)
        } else {
            HybridAutoPlay.safeAreaInsetsListeners[moduleName] = [callback]
        }

        return {
            HybridAutoPlay.safeAreaInsetsListeners[moduleName]?.removeAll {
                $0 as AnyObject === callback as AnyObject
            }
            if HybridAutoPlay.safeAreaInsetsListeners[moduleName]?.isEmpty
                ?? false
            {
                HybridAutoPlay.safeAreaInsetsListeners.removeValue(
                    forKey: moduleName
                )
            }
        }
    }

    func presentTemplate(templateId: String) throws
        -> NitroModules.Promise<Void>
    {
        return Promise.async {
            return try await RootModule.withTemplateAndInterfaceController(
                templateId: templateId
            ) { template, interfaceController in
                let _ = try await interfaceController.presentTemplate(
                    template,
                    animated: true
                )
            }
        }
    }
    
    func dismissTemplate(animate: Bool?) throws
        -> NitroModules.Promise<Void>
    {
        return Promise.async {
            return try await RootModule.withInterfaceController() { interfaceController in
                let _ = try await interfaceController.dismissTemplate(
                    animated: animate ?? true
                )
            }
        }
    }

    // MARK: set/push/pop templates
    func setRootTemplate(templateId: String) throws -> Promise<Void> {
        return Promise.async {
            try await RootModule.withSceneTemplateAndInterfaceController(
                templateId: templateId
            ) { template, scene, interfaceController in
                if template is CPMapTemplate {
                    await MainActor.run {
                        scene.initRootView()
                    }
                }

                let _ = try await interfaceController.setRootTemplate(
                    template,
                    animated: false
                )
            }
        }
    }

    func pushTemplate(templateId: String) throws
        -> NitroModules.Promise<Void>
    {
        return Promise.async {
            return try await RootModule.withTemplateAndInterfaceController(
                templateId: templateId
            ) { template, interfaceController in
                let _ = try await interfaceController.pushTemplate(
                    template,
                    animated: true
                )
            }
        }
    }

    func popTemplate(animate: Bool?) throws -> NitroModules.Promise<Void> {
        return Promise.async {
            return try await RootModule.withInterfaceController {
                interfaceController in
                
                let hasPresentedTemplate = await interfaceController.hasPresentedTemplate()
                if (hasPresentedTemplate) {
                    let presentedTemplateId = try await interfaceController.dismissTemplate(animated: animate ?? true)
                    if (presentedTemplateId != nil) {
                        HybridAutoPlay.removeListeners(templateId: presentedTemplateId!)
                    }
                    return
                }
                
                guard
                    let templateId = try await interfaceController.popTemplate(
                        animated: true
                    )
                else { return }
                HybridAutoPlay.removeListeners(templateId: templateId)
            }
        }
    }

    func popToRootTemplate(animate: Bool?) throws -> NitroModules.Promise<Void> {
        return Promise.async {
            try await RootModule.withInterfaceController {
                interfaceController in
                
                let hasPresentedTemplate = await interfaceController.hasPresentedTemplate()
                if (hasPresentedTemplate) {
                    let presentedTemplateId = try await interfaceController.dismissTemplate(animated: animate ?? true)
                    if (presentedTemplateId != nil) {
                        HybridAutoPlay.removeListeners(templateId: presentedTemplateId!)
                    }
                }
                
                let templateIds =
                    try await interfaceController.popToRootTemplate(
                        animated: animate ?? true
                    )
                for templateId in templateIds {
                    HybridAutoPlay.removeListeners(templateId: templateId)
                }
            }
        }
    }

    func popToTemplate(templateId: String) throws -> Promise<Void> {
        return Promise.async {
            return try await RootModule.withInterfaceController {
                interfaceController in

                let templateIds = try await interfaceController.popToTemplate(
                    templateId: templateId,
                    animated: true
                )
                templateIds.forEach { templateId in
                    HybridAutoPlay.removeListeners(templateId: templateId)
                }
            }
        }
    }

    // MARK: generic template updates
    func setTemplateHeaderActions(templateId: String, headerActions: [NitroAction]?) throws
    {
        try RootModule.withTemplate(templateId: templateId) {
            template in
            template.barButtons = headerActions
            template.setBarButtons()
        }
    }

    // MARK: events
    static func emit(event: EventName) {
        HybridAutoPlay.listeners[event]?.values.forEach {
            $0()
        }
    }

    static func emitRenderState(mapTemplateId: String, state: VisibilityState) {
        HybridAutoPlay.renderStateListeners[mapTemplateId]?.forEach {
            callback in
            callback(state)
        }
    }

    static func emitSafeAreaInsets(
        moduleName: String,
        safeAreaInsets: UIEdgeInsets
    ) {
        let insets = SafeAreaInsets(
            top: safeAreaInsets.top,
            left: safeAreaInsets.left,
            bottom: safeAreaInsets.bottom,
            right: safeAreaInsets.right,
            isLegacyLayout: nil
        )
        HybridAutoPlay.safeAreaInsetsListeners[moduleName]?.forEach {
            callback in callback(insets)
        }
    }

    static func removeListeners(templateId: String) {
        HybridAutoPlay.renderStateListeners.removeValue(forKey: templateId)
        HybridAutoPlay.safeAreaInsetsListeners.removeValue(forKey: templateId)
    }
}

File: packages/react-native-autoplay/android/src/main/java/com/margelo/nitro/at/g4rb4g3/autoplay/hybrid/HybridAutoPlay.kt
package com.margelo.nitro.at.g4rb4g3.autoplay.hybrid

import com.margelo.nitro.at.g4rb4g3.autoplay.ActivityRenderStateProvider
import com.margelo.nitro.at.g4rb4g3.autoplay.AndroidAutoScreen
import com.margelo.nitro.at.g4rb4g3.autoplay.AndroidAutoSession
import com.margelo.nitro.at.g4rb4g3.autoplay.VirtualRenderer
import com.margelo.nitro.at.g4rb4g3.autoplay.template.AndroidAutoTemplate
import com.margelo.nitro.at.g4rb4g3.autoplay.utils.ThreadUtil
import com.margelo.nitro.core.Promise

class HybridAutoPlay : HybridHybridAutoPlaySpec() {
    override fun addListener(
        eventType: EventName, callback: () -> Unit
    ): () -> Unit {
        val callbacks = listeners.getOrPut(eventType) { mutableListOf() }
        callbacks.add(callback)

        if (eventType == EventName.DIDCONNECT && AndroidAutoSession.Companion.getIsConnected()) {
            callback()
        }

        return {
            listeners[eventType]?.removeAll { it === callback }
        }
    }

    override fun isConnected(): Boolean {
        return AndroidAutoSession.getIsConnected()
    }

    override fun addListenerRenderState(
        moduleName: String, callback: (VisibilityState) -> Unit
    ): () -> Unit {
        val callbacks = renderStateListeners.getOrPut(moduleName) {
            mutableListOf()
        }
        callbacks.add(callback)

        if (moduleName == "main") {
            callback(ActivityRenderStateProvider.currentState)
        } else {
            AndroidAutoSession.Companion.getState(moduleName)?.let {
                callback(it)
            }
        }

        return {
            renderStateListeners[moduleName]?.let {
                it.remove(callback)
                if (it.isEmpty()) {
                    renderStateListeners.remove(moduleName)
                }
            }
        }
    }

    override fun addSafeAreaInsetsListener(
        moduleName: String, callback: (SafeAreaInsets) -> Unit
    ): () -> Unit {
        val callbacks = safeAreaInsetsListeners.getOrPut(moduleName) {
            mutableListOf()
        }
        callbacks.add(callback)

        return {
            safeAreaInsetsListeners[moduleName]?.let {
                it.remove(callback)
                if (it.isEmpty()) {
                    safeAreaInsetsListeners.remove(moduleName)
                }
            }
        }
    }

    override fun setTemplateHeaderActions(
        templateId: String, headerActions: Array<NitroAction>?
    ): Promise<Unit> {
        return Promise.async {
            val template = AndroidAutoTemplate.Companion.getTemplate(templateId)
                ?: throw IllegalArgumentException(
                    "setTemplateHeaderActions failed, template $templateId not found"
                )
            template.setTemplateHeaderActions(headerActions)
        }
    }

    override fun setRootTemplate(templateId: String): Promise<Unit> {
        return Promise.Companion.async {
            val template = AndroidAutoTemplate.Companion.getTemplate(templateId)
                ?: throw IllegalArgumentException("setRootTemplate failed, $templateId template not found")

            if (template.isRenderTemplate) {
                val screen = AndroidAutoScreen.Companion.getScreen(templateId)
                    ?: throw IllegalArgumentException("setRootTemplate failed, $templateId screen not found")
                val carContext = AndroidAutoSession.Companion.getCarContext(templateId)
                    ?: throw IllegalArgumentException("setRootTemplate failed, carContext for $templateId template not found")

                screen.applyConfigUpdate(invalidate = true)

                if (!VirtualRenderer.Companion.hasRenderer(templateId)) {
                    val result = ThreadUtil.postOnUiAndAwait {
                        VirtualRenderer(carContext, templateId)
                    }
                    if (result.isFailure) {
                        throw result.exceptionOrNull()
                            ?: UnknownError("unknown error initializing the virtual screen")
                    }
                }
            } else {
                val screenManager = AndroidAutoScreen.Companion.getScreenManager()
                    ?: throw IllegalArgumentException("setRootTemplate failed, screenManager not found")
                val carContext =
                    AndroidAutoSession.Companion.getCarContext(AndroidAutoSession.Companion.ROOT_SESSION)
                        ?: throw IllegalArgumentException("setRootTemplate failed, carContext for $templateId template not found")

                val result = ThreadUtil.postOnUiAndAwait {
                    screenManager.popToRoot()
                    val previousRootScreen = screenManager.top

                    val screen = AndroidAutoScreen(carContext, templateId, template.parse())
                    screenManager.push(screen)

                    screenManager.remove(previousRootScreen)
                }

                if (result.isFailure) {
                    throw result.exceptionOrNull()
                        ?: UnknownError("unknown error setting root template")
                }
            }
        }
    }

    override fun pushTemplate(templateId: String): Promise<Unit> {
        return Promise.Companion.async {
            val context = AndroidAutoSession.Companion.getRootContext()
                ?: throw IllegalArgumentException("pushTemplate failed, carContext not found")
            val template = AndroidAutoTemplate.Companion.getTemplate(templateId)
                ?: throw IllegalArgumentException("pushTemplate failed, template $templateId not found")
            val screenManager = AndroidAutoScreen.Companion.getScreenManager()
                ?: throw IllegalArgumentException("pushTemplate failed, screenManager not found")

            val result = ThreadUtil.postOnUiAndAwait {
                val screen = AndroidAutoScreen(context, templateId, template.parse())
                screenManager.push(screen)
            }

            if (result.isFailure) {
                throw result.exceptionOrNull() ?: UnknownError("unknown error pushing template")
            }
        }
    }

    override fun popTemplate(animate: Boolean?): Promise<Unit> {
        return Promise.Companion.async {
            val screenManager = AndroidAutoScreen.Companion.getScreenManager()
                ?: throw IllegalArgumentException("popTemplate failed, screenManager not found")
            if (screenManager.stackSize == 0) {
                return@async
            }

            val result = ThreadUtil.postOnUiAndAwait {
                screenManager.pop()
            }

            if (result.isFailure) {
                throw result.exceptionOrNull() ?: UnknownError("unknown error popping template")
            }
        }
    }

    override fun popToRootTemplate(animate: Boolean?): Promise<Unit> {
        return Promise.Companion.async {
            val screenManager = AndroidAutoScreen.Companion.getScreenManager()
                ?: throw IllegalArgumentException("popToRootTemplate failed, screenManager not found")
            if (screenManager.stackSize == 0) {
                return@async
            }

            val result = ThreadUtil.postOnUiAndAwait {
                screenManager.popToRoot()
            }

            if (result.isFailure) {
                throw result.exceptionOrNull() ?: UnknownError("unknown error popping template")
            }
        }
    }

    override fun popToTemplate(templateId: String, animate: Boolean?): Promise<Unit> {
        return Promise.Companion.async {
            val screenManager = AndroidAutoScreen.Companion.getScreenManager()
                ?: throw IllegalArgumentException("pushTemplate failed, screenManager not found")
            if (screenManager.stackSize == 0) {
                return@async
            }

            val result = ThreadUtil.postOnUiAndAwait {
                screenManager.popTo(templateId)
            }

            if (result.isFailure) {
                throw result.exceptionOrNull()
                    ?: UnknownError("unknown error popping template $templateId")
            }
        }
    }

    companion object {
        const val TAG = "HybridAutoPlay"

        private val listeners = mutableMapOf<EventName, MutableList<() -> Unit>>()

        private val renderStateListeners =
            mutableMapOf<String, MutableList<(VisibilityState) -> Unit>>()
        private val safeAreaInsetsListeners =
            mutableMapOf<String, MutableList<(SafeAreaInsets) -> Unit>>()


        fun removeListeners(templateId: String) {
            renderStateListeners.remove(templateId)
            safeAreaInsetsListeners.remove(templateId)
        }

        fun emit(event: EventName) {
            listeners[event]?.forEach { it() }
        }

        fun emitRenderState(moduleName: String, state: VisibilityState) {
            renderStateListeners[moduleName]?.forEach {
                it(state)
            }
        }

        fun emitSafeAreaInsets(
            moduleName: String,
            top: Double,
            bottom: Double,
            left: Double,
            right: Double,
            isLegacyLayout: Boolean
        ) {
            val insets = SafeAreaInsets(
                top = top,
                bottom = bottom,
                left = left,
                right = right,
                isLegacyLayout = isLegacyLayout
            )
            safeAreaInsetsListeners[moduleName]?.forEach {
                it(insets)
            }
        }
    }
}